source/base/quadrature_lib.cc

```diff
@@ -1345,6 +1345,165 @@ QGaussLobattoChebyshev<dim>::QGaussLobattoChebyshev (const unsigned int n)
   Quadrature<dim> (QGaussLobattoChebyshev<1>(n))
 {}
 
+
+
+template<int dim>
+QSimplex<dim>::QSimplex(const Quadrature<dim> &quad)
+{
+  std::vector<Point<dim> > qpoints;
+  std::vector<double > weights;
+
+  for (unsigned int i=0; i<quad.size(); ++i)
+    {
+      double r=0;
+      for (unsigned int d=0; d<dim; ++d)
+        r += quad.point(i)[d];
+      if (r <= 1)
+        {
+          this->quadrature_points.push_back(quad.point(i));
+          this->weights.push_back(quad.weight(i));
+        }
+    }
+}
+
+
+
+template<int dim>
+Quadrature<dim>
+QSimplex<dim>::compute_affine_transformation(const std::array<Point<dim>, dim+1>& vertices) const
+{
+  unsigned int i=0;
+  Tensor<2,dim> B;
+  for (unsigned int d=0; d<dim; ++d)
+    B[d] = vertices[d+1]-vertices[0];
+
+  B = transpose(B);
+  auto J = std::abs(determinant(B));
+
+  // if the determinant is zero, we return an empty quadrature
+  if (J < 1e-12)
+    return Quadrature<dim>();
+
+  std::vector<Point<dim> > qp(this->size());
+  std::vector<double> w(this->size());
+
+  for (unsigned int i=0; i<this->size(); ++i)
+    {
+      qp[i] = Point<dim>(vertices[0]+B*this->point(i));
+      w[i] = J*this->weight(i);
+    }
+
+  return Quadrature<dim>(qp, w);
+}
+
+
+
+QTrianglePolar::QTrianglePolar(const Quadrature<1> &radial_quadrature,
+                               const Quadrature<1> &angular_quadrature) :
+  QSimplex<2>(Quadrature<2>())
+{
+  QAnisotropic<2> base(radial_quadrature, angular_quadrature);
+  this->quadrature_points.resize(base.size());
+  this->weights.resize(base.size());
+  for (unsigned int i=0; i<base.size(); ++i)
+    {
+      const auto &q = base.point(i);
+      const auto &w = base.weight(i);
+
+      const auto &xhat = q[0];
+      const auto &yhat = q[1];
```

Comment:
for all of these scalar objects, I think dropping the reference would make the compiler's life easier. That's because the rhs operation returns  temporary, and you force the compiler to extend destruction of the temporary to the end of the scope. same for `pi` below.